### ISSUE TITLE
cub::BlockStore::Store doesn't compile if OutputIteratorT::value_type != T

### ISSUE THREAD

**Issue by egaburov** [NVlabs/cub#69]
OutputIteartorT must be templated in
https://github.com/NVlabs/cub/blob/1.5.5/cub/block/block_store.cuh#L947
e.g.  if it is discard_iterator 

cub:BlockStore must be templated on type T, rather than iterator

**Comment by dumerrill** [comment]
I'm not sure what you mean.  Type T inside BlockStore is literally defined as the value_type of OutputIteratorT ( https://github.com/NVlabs/cub/blob/1.5.5/cub/block/block_store.cuh#L518):

```
    // Data type of iterator
    typedef typename std::iterator_traits<OutputIteratorT>::value_type T;
```

Perhaps there is some issue with your iterator traits?

**Comment by dumerrill** [comment]
Perhaps you are attempting to write thread data of some other value type than T = std::iterator_traits<OutputIteratorT>::value_type, e.g., an array of items that are convertible/coercible to T.  In that case, you can simply construct an array of T yourself, doing the conversion prior to the call to Store().  (The compiler will optimize most, if not all, of it away.)

**Comment by egaburov** [comment]
output_iterator::value_type may be void, and accepts anything, e.g.  {thrust|boost}::discard_iterator. The correct design would be to template on output iterator argument on BlockStore::Store method. There will be only one operation, after all transformations, that writes data to the output iterator.

**Comment by dumerrill** [comment]
Actually, I believe the value_type for any output iterator is expected to be void.  Definitely a bug then; will update interface (which unfortunately changes the class template interface)

**Comment by dumerrill** [comment]
Fixed in the above commit for v1.6.3
